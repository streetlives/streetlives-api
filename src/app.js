import express from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import morgan from 'morgan';
import awsServerlessExpressMiddleware from 'aws-serverless-express/middleware';
import setupRoutes from './routes';
// import models from './models';

const app = express();

app.use(morgan('dev'));

app.use(cors());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));
app.use(awsServerlessExpressMiddleware.eventContext());

app.use(bodyParser.json());

setupRoutes(app);

// TODO: Decide what to do about this (at least needs fixing for tests).
// models.sequelize.sync();

export default app;
