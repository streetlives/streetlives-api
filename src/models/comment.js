module.exports = (sequelize, DataTypes) => {
  const Comment = sequelize.define('Comment', {
    id: {
      type: DataTypes.UUID,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV4,
    },
    content: DataTypes.TEXT,
    posted_by: DataTypes.TEXT,
    contact_info: DataTypes.TEXT,
  }, {
    underscored: true,
    underscoredAll: true,
  });

  Comment.associate = (models) => {
    Comment.belongsTo(models.Location);
    Comment.belongsTo(models.ServiceAtLocation);
    Comment.belongsTo(models.Comment, { as: 'ReplyTo', foreignKey: 'reply_to_id' });
    Comment.hasMany(models.Comment, { as: 'Replies', foreignKey: 'reply_to_id' });
  };

  return Comment;
};
