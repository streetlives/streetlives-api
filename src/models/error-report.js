module.exports = (sequelize, DataTypes) => {
  const ErrorReport = sequelize.define('ErrorReport', {
    id: {
      type: DataTypes.UUID,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV4,
    },
    general_location_error: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
    },
    services: {
      // May need to create an index for this datatype (array of UUIDs) for look-ups
      // https://coderwall.com/p/1b5eyq/index-for-uuid-array-data-type

      type: DataTypes.ARRAY(DataTypes.UUID),
      allowNull: false,
      defaultValue: [],
    },
    content: DataTypes.TEXT,
  }, {
    underscored: true,
    underscoredAll: true,
  });

  ErrorReport.associate = (models) => {
    ErrorReport.belongsTo(models.Location);
  };

  ErrorReport.findAllForLocation = (locationId, { attributes, order }) => ErrorReport.findAll({
    where: {
      location_id: locationId,
    },
    attributes,
    order,
  });

  return ErrorReport;
};
