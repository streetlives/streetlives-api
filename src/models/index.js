/* eslint-disable global-require */
/* eslint-disable import/no-dynamic-require */
import fs from 'fs';
import path from 'path';
import Sequelize, { DataTypes, Op } from 'sequelize';
import config from '../config';

const basename = path.basename(__filename);
const db = {};

const sequelize = new Sequelize(
  config.db.database,
  config.db.username,
  config.db.password,
  config.db.options,
);

fs
  .readdirSync(__dirname)
  .filter(file => (file.indexOf('.') !== 0) && (file !== basename) && (file.slice(-3) === '.js'))
  .forEach((file) => {
    const model = require(`./${file}`)(sequelize, DataTypes, Op);
    db[model.name] = model;
  });

Object.keys(db).forEach((modelName) => {
  if (db[modelName].associate) {
    db[modelName].associate(db);
  }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;
