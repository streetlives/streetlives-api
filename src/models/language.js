module.exports = (sequelize, DataTypes) => {
  const Language = sequelize.define('Language', {
    id: {
      type: DataTypes.UUID,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV4,
    },
    language: DataTypes.TEXT,
    name: DataTypes.TEXT,
  }, {
    underscored: true,
    underscoredAll: true,
  });

  Language.associate = (models) => {
    Language.belongsToMany(models.Location, { through: models.LocationLanguages });
    Language.belongsToMany(models.Service, { through: models.ServiceLanguages });
  };

  return Language;
};
