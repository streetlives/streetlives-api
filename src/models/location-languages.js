module.exports = (sequelize, DataTypes) => {
  const LocationLanguages = sequelize.define('LocationLanguages', {
    id: {
      type: DataTypes.UUID,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV4,
    },
  }, {
    underscored: true,
    underscoredAll: true,
  });

  return LocationLanguages;
};
