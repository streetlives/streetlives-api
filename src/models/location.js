module.exports = (sequelize, DataTypes) => {
  const Location = sequelize.define('Location', {
    id: {
      type: DataTypes.UUID,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV4,
    },
    name: DataTypes.TEXT,
    description: DataTypes.TEXT,
    transportation: DataTypes.TEXT,
    position: DataTypes.GEOMETRY,
    additional_info: DataTypes.TEXT,
  }, {
    underscored: true,
    underscoredAll: true,
  });

  Location.associate = (models) => {
    Location.belongsTo(models.Organization);
    Location.belongsToMany(models.Service, { through: models.ServiceAtLocation });
    Location.belongsToMany(models.Language, { through: models.LocationLanguages });
    Location.hasMany(models.PhysicalAddress);
    Location.hasMany(models.Phone);
    Location.hasMany(models.RegularSchedule);
    Location.hasMany(models.HolidaySchedule);
    Location.hasMany(models.AccessibilityForDisabilities);
    Location.hasMany(models.Comment);
  };

  Location.findAllInArea = (position, radius, filterParameters) => {
    const { searchString, taxonomyId } = filterParameters;

    const distance = sequelize.fn(
      'ST_Distance_Sphere',
      sequelize.literal('position'),
      sequelize.literal(`ST_GeomFromGeoJSON('${JSON.stringify(position)}')`),
    );

    const conditions = [];
    conditions.push(sequelize.where(distance, { $lte: radius }));

    if (searchString) {
      const fuzzySearchString = `%${searchString}%`;
      conditions.push(sequelize.or(
        { name: { [sequelize.Op.iLike]: fuzzySearchString } },
        { '$Organization.name$': { [sequelize.Op.iLike]: fuzzySearchString } },
        { '$Services.name$': { [sequelize.Op.iLike]: fuzzySearchString } },
        { '$Services.Taxonomies.name$': { [sequelize.Op.iLike]: fuzzySearchString } },
      ));
    }

    if (taxonomyId) {
      conditions.push({ '$Services.Taxonomies.id$': taxonomyId });
    }

    return Location.findAll({
      where: sequelize.and(...conditions),
      include: [
        sequelize.models.Organization,
        sequelize.models.Phone,
        sequelize.models.PhysicalAddress,
        {
          model: sequelize.models.Service,
          include: sequelize.models.Taxonomy,
        },
      ],
    });
  };

  return Location;
};
