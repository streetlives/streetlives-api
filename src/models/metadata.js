module.exports = (sequelize, DataTypes) => {
  const actionTypes = {
    create: 'create',
    update: 'update',
    delete: 'delete',
  };

  const Metadata = sequelize.define('Metadata', {
    id: {
      type: DataTypes.UUID,
      primaryKey: true,
      defaultValue: DataTypes.UUIDV4,
    },
    resource_id: {
      type: DataTypes.UUID,
      allowNull: false,
    },
    resource_table: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    last_action_date: {
      type: DataTypes.DATE,
      allowNull: false,
    },
    last_action_type: {
      type: DataTypes.ENUM,
      values: Object.keys(actionTypes).map( k => actionTypes[k] ),
      allowNull: false,
    },
    field_name: DataTypes.TEXT,
    previous_value: DataTypes.TEXT,
    replacement_value: DataTypes.TEXT,
    updated_by: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
  }, {
    tableName: 'metadata',
    underscored: true,
    underscoredAll: true,
  });

  Metadata.actionTypes = actionTypes;

  Metadata.associate = (models) => {
    Object.keys(models).forEach((modelKey) => {
      const model = models[modelKey];

      if (model !== models.Metadata) {
        Metadata.belongsTo(model, { foreignKey: 'resource_id', constraints: false });
        model.hasMany(Metadata, {
          foreignKey: 'resource_id',
          constraints: false,
          scope: {
            resource_table: model.tableName,
          },
        });
      }
    });
  };

  Metadata.getLastUpdateDatesForResourceFields = resourceId => Metadata.findAll({
    attributes: [
      'field_name',
      [sequelize.fn('MAX', sequelize.col('last_action_date')), 'last_action_date'],
    ],
    where: { resource_id: resourceId },
    group: 'field_name',
  });

  Metadata.getLatestUpdateDateForQuery = where => Metadata.max('last_action_date', { where });

  Metadata.getLatestUpdateDateForResource = resourceId => Metadata.getLatestUpdateDateForQuery({
    resource_id: resourceId,
  });

  Metadata.getLatestUpdateDateForResources = resourceIds => Metadata.getLatestUpdateDateForQuery({
    resource_id: { [sequelize.Op.in]: resourceIds },
  });

  return Metadata;
};
