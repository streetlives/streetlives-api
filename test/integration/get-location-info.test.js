import request from 'supertest';
import app from '../../src';
import models from '../../src/models';

describe('get location info', () => {
  let location;

  const setupData = () => models.Organization.create({
    name: 'The Test Org',
    description: 'An organization meant for testing purposes.',
    url: 'www.streetlives.com',
  })
    .then(organization => models.Location.create({
      organization_id: organization.id,
      name: 'Some kind of center',
      PhysicalAddresses: [{
        address_1: '123 West 30th Street',
        city: 'New York',
        state_province: 'NY',
        postal_code: '10001',
        country: 'United States',
      }],
      Phones: [{
        number: '212.121.0123',
      }],
      Services: [{
        organization_id: organization.id,
        name: 'A specific offering',
        Taxonomies: [{
          name: 'Shelter',
        }],
      }],
      Comments: [{
        content: 'Test comment',
        posted_by: 'Test poster',
      }],
    }, {
      include: [
        { model: models.PhysicalAddress },
        { model: models.Phone },
        { model: models.Comment },
        { model: models.Service, include: [{ model: models.Taxonomy }] },
      ],
    }))
    .then((newLocation) => {
      location = newLocation;
    });

  beforeAll(() => models.sequelize.sync({ force: true }).then(setupData));
  afterAll(() => {
    models.sequelize.close();
    app.server.close();
  });

  const stripTimestampsAndIds = obj => Object.keys(obj).reduce((currStrippedObj, key) => {
    const value = obj[key];

    if (key === 'id' || key.includes('_id') || key.includes('_at')) {
      return currStrippedObj;
    }

    if (value && typeof value === 'object') {
      return { ...currStrippedObj, [key]: stripTimestampsAndIds(value) };
    }
    return { ...currStrippedObj, [key]: value };
  }, {});

  it('should get all relevant fields for a given location', () =>
    request(app)
      .get(`/locations/${location.id}`)
      .expect(200)
      .then((res) => {
        const strippedFields = stripTimestampsAndIds(res.body);
        expect(strippedFields).toMatchSnapshot();
      }));

  it('should return a 404 status code when the given location ID isn\'t found', () => {
    const nonExistentLocationId = '11111111-1111-1111-1111-111111111111';

    return request(app)
      .get(`/locations/${nonExistentLocationId}`)
      .expect(404);
  });
});
