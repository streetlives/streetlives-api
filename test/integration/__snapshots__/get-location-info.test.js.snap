// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get location info should get all relevant fields for a given location 1`] = `
Object {
  "Comments": Object {
    "0": Object {
      "content": "Test comment",
      "posted_by": "Test poster",
    },
  },
  "Organization": Object {
    "description": "An organization meant for testing purposes.",
    "email": null,
    "name": "The Test Org",
    "url": "www.streetlives.com",
  },
  "Phones": Object {
    "0": Object {
      "description": null,
      "extension": null,
      "language": null,
      "number": "212.121.0123",
      "type": null,
    },
  },
  "Services": Object {
    "0": Object {
      "Languages": Object {},
      "RegularSchedules": Object {},
      "ServiceAtLocation": Object {
        "description": null,
      },
      "Taxonomies": Object {
        "0": Object {
          "ServiceTaxonomy": Object {},
          "name": "Shelter",
          "parent_name": null,
        },
      },
      "description": null,
      "email": null,
      "fees": null,
      "interpretation_services": null,
      "metadata": Object {
        "service": Object {},
      },
      "name": "A specific offering",
      "url": null,
    },
  },
  "address": Object {
    "city": "New York",
    "country": "United States",
    "postalCode": "10001",
    "region": null,
    "state": "NY",
    "street": "123 West 30th Street",
  },
  "description": null,
  "metadata": Object {
    "address": Object {},
    "location": Object {},
    "organization": Object {},
  },
  "name": "Some kind of center",
  "position": null,
  "transportation": null,
}
`;
